bundles/org.openhab.binding.somfymylink/src/main/java/org/openhab/binding/somfymylink/internal/handler/SomfyMyLinkBridgeHandler.java

```diff
@@ -0,0 +1,504 @@
+                    int readCount;
+                    String message = "";
+
+                    // while we are getting data ...
+                    while (((readCount = in.read(readBuff)) != -1)) {
+                        message += new String(readBuff, 0, readCount);
+
+                        // try and parse the message
+                        try {
+                            T response = parseResponse(message, responseType);
+                            logger.debug("Got full message: {}", message);
+                            future.complete(response);
+                            return;
+                        } catch (SomfyMyLinkException e) {
+                            // trouble parsing the message, probably incomplete
+                            // if its been over 1 min waiting for a correct and full response then abort
+                            elapsedTime = (new Date()).getTime() - startTime;
+                            if (elapsedTime >= 60000) {
+                                future.completeExceptionally(new SomfyMyLinkException("Timeout waiting for reply from mylink"));
+                                return;
+                            }
+                            Thread.sleep(250); // pause for 250ms
+                        }
+                    }
+                }
+
+                // only if we didn't already get a response give time for mylink to process
+                Thread.sleep(CONNECTION_DELAY);
+
+                future.complete(null);
+            } catch (SocketTimeoutException e) {
+                logger.info("Timeout sending command to mylink: {} Message: {}", command, e.getMessage());
+                future.completeExceptionally(new SomfyMyLinkException("Timeout sending command to mylink", e));
+            } catch (IOException e) {
+                logger.info("Problem sending command to mylink: {} Message: {}", command, e.getMessage());
+                future.completeExceptionally(new SomfyMyLinkException("Problem sending command to mylink", e));
+            } catch (InterruptedException e) {
+                logger.debug("Interrupted while waiting after sending command to mylink: {} Message: {}", command, e.getMessage());
+                future.complete(null);
+            }
+        });
+        return future;
+    }
+
+    private <T extends SomfyMyLinkResponseBase> T parseResponse(String message, Class<T> responseType) {
+        JsonParser parser = new JsonParser();
+        JsonObject jsonObj = parser.parse(message).getAsJsonObject();
+
+        if (jsonObj != null && jsonObj.has("error")) {
+            SomfyMyLinkErrorResponse errorResponse = gson.fromJson(message, SomfyMyLinkErrorResponse.class);
+
+            logger.info("Error parsing mylink response: {}", errorResponse.error.message);
+            throw new SomfyMyLinkException("Incomplete message.");
+        }
+
+        return gson.fromJson(message, responseType);
+    }
```

Comment:
I've noticed that you are trying to parse the json message each time you receive a chunk of the message from the reader. Perhaps it would be better to just give the json parser the Reader instance to begin with so that you don't need to handle buffering or partial messages.
```suggestion
    private <T extends SomfyMyLinkResponseBase> T parseResponse(Reader reader, Class<T> responseType) {
        JsonParser parser = new JsonParser();
        JsonElement jsonElement = parser.parse(reader);
        JsonObject jsonObj = jsonElement.getAsJsonObject();

        if (jsonObj != null && jsonObj.has("error")) {
            SomfyMyLinkErrorResponse errorResponse = gson.fromJson(jsonElement, SomfyMyLinkErrorResponse.class);

            logger.info("Error parsing mylink response: {}", errorResponse.error.message);
            throw new SomfyMyLinkException("Incomplete message.");
        }

        return gson.fromJson(jsonElement, responseType);
    }
```

Reply:
Unfortunately the somfy mylink doesnt sent newline at the of response messages, so the direct json parsing from the BufferedReader fails (times out).

If you have any further thoughts on how i could do this better im all ears, but right now reading char[] from the socket and trying to parse seems like my only option?

Reply:
What if you try adding:
```
parser.setLenient(true);
```
before parsing? Does that work?

Reply:
> Unfortunately the somfy mylink doesnt sent newline at the of response messages, so the direct json parsing from the BufferedReader fails (times out).

I'm not sure what you mean. A newline isn't required for the BufferedReader to operate, the `readLine()` method is simply a convenience function, the BufferedReader doesn't only read a single line at a time, it fills its entire buffer.

Now if you instead mean that the somfy mylink doesn't close the stream after it finishes writing the json then I can understand why it would fail and can come up with a solution. The only problem is that such a case should cause your current solution to fail as well so I might be misunderstanding the problem that you are experiencing. 

I know that there is way to get gson reading directly from the Reader to work, but I need to get a better understanding of the problem you are experiencing.

Reply:
I tried the readLine() option but it blocks and then times out. Presumably (or so i thought) due to no newline. But as you point out it might be due to no stream closure.

I thought it might be new line after some reading things like [this](https://stackoverflow.com/questions/19194813/how-to-read-a-line-which-doesnt-end-with-new-line-or-carriage-return-character).

The current solution reads until there is no more data, so it probably doesnt care about stream closure? I can verify this works from testing it with a real mylink device.

Reply:
The strange thing is that if you passed the buffered reader directly into gson, it should also read until the end of stream like you do now. But you said that didn't work so I've been very curious about what is going on.

If it was actually a stream closure issue then I would expect that your current solution wouldn't work either. But since your current solution works I'm curious if there is something else going on that is preventing you from passing the reader directly to gson.

Reply:
Ok i did a bit more testing with the following code:
```
if (in.ready()) {
    int available = socket.getInputStream().available();
    data = in.readLine();
}
```
In this case I saw 738 bytes available, however the readLine threw a SocketTimeoutException. 

Then i tried this:
```
if (in.ready()) {
    int available = socket.getInputStream().available();
    char[] readBuff = new char[1024];
    long size = in.read(readBuff);
    int available2 = socket.getInputStream().available();
    long size2 = in.read(readBuff);
}
```
It showed 738 bytes in `available` and read them into `readBuff` (see below). Then `available2` showed `0` as expected and `read` threw a `SocketTimeoutException`.

The `readBuff` showed (snipped to the last few bytes):
```
735:3
736:5
737:}
738:
```

Looking at the docs for readLine it says:
```
The java.io.BufferedReader.readline() method read a line of text. A line is considered to be terminated by any one of a line feed ('\n'), a carriage return ('\r'), or a carriage return followed immediately by a linefeed.
```
From the above it looks like the myLink isn't sending a newline or closing the socket after it's finished sending.